### ISSUE TITLE
GPU Scaling overhead on Summit

### ISSUE THREAD

**Issue by lee212** [radical-collaboration/ExTASY#1]
1hr non-productive time is reported on starting `2^{12}` GPU tasks on Summit. 

TODO items are here to assess the high overhead time:

- [x] make a copy of session/sandbox data
- [x] validate profile data to check gaps between job submission and exec start
- [x] run a prte test
- [x] `2^{10}` test run and measure overhead
- [x] `2^{11}` test run and measure overhead
- [x] verify the current RCT stack versions
- [x] look through debugging messages in log files to detect abnormal behavior
- [x] consider other factors, e.g. shared file system delay
- [ ] add a benchmark result section to readthedocs

**Comment by lee212** [comment]
[re.session.login4.eh22.018340.0007.tar.gz](https://github.com/radical-collaboration/ExTASY/files/4360889/re.session.login4.eh22.018340.0007.tar.gz)
[re.session.login4.eh22.018340.0007.zip](https://github.com/radical-collaboration/ExTASY/files/4360182/re.session.login4.eh22.018340.0007.zip)

This session data is not about scaling issue but a prte failure.
~~Also, there is no sandbox.~~

**Comment by lee212** [comment]
[re.session.login5.eh22.018316.0009.tar.gz](https://github.com/radical-collaboration/ExTASY/files/4361904/re.session.login5.eh22.018316.0009.tar.gz)

sandbox for 1hr overhead prte run

**Comment by mturilli** [comment]
I am not sure this applies but I noticed the following with some of my own runs. In my case, I was running RCT stack from one of our workstations towards Frontera. I had to stage 5 files, one of which of 1GB. For causes still under investigation, the staging took a long time, approaching indeed 1 hour. Could this be the case also for this ticket?

**Comment by mturilli** [comment]
@andre-merzky this has been analyzed for a while now but no conclusive answer was found. Would you mind to give a look at the session and see whether you can confirm that the observed overhead is indeed due to data staging?

**Comment by andre-merzky** [comment]
One of the task STDERR files was hiding this:
```
unit.000044/STDERR:prun: Error: unknown option "--hnp"
```
That is the only task which finishes (with that error), all other tasks seem not to complete within pilot runtime - but they seem to start since the `md.log` files are getting populated.  I assume `unit.000044` fails because of the `module load ibm-wml-ce/1.6.1-3` it has in the `pre_exec`, which seems to mess with the PRTE module.  That is tricky, and we should ask ORNL about how to handle that case.   One work around would be to create a shell wrapper for the task which does the module load.

Otherwise the bootstrapper needs 18 seconds, which seems decent, and the first task gets started after 110 more seconds,  which looks ok (EnTK waits for the pilot to be ACTIVE before event submitting the tasks).  The tasks all get started within 6 seconds - not terribly fast, but not terribly slow either.

Does that help?

**Comment by lee212** [comment]
Thanks @andre-merzky for the checking the session!

However, the session data to investigate are located on Summit,  they are not attached to github due to the size. Sorry for the confusion.

They are under `/gpfs/alpine/proj-shared/bip191/scaling`

The settings for the sessions are:
**400 nodes=2400GPUs=3030s startup: re.session.login5.eh22.018314.0016**
10 nodes: re.session.login1.eh22.018313.0003
10 nodes prte: re.session.login2.eh22.018340.0001

The 2400GPUs session shows about 1hr overhead which you may want to check out.

**Comment by andre-merzky** [comment]
Alas:
```sh
$ cd /gpfs/alpine/proj-shared/bip191/scaling
-bash: cd: /gpfs/alpine/proj-shared/`/scaling: Permission denied
```
I am not a member of `BIP191`

**Comment by lee212** [comment]
Oops.  Could you try again?
`/gpfs/alpine/world-shared/bip191/scaling `

**Comment by andre-merzky** [comment]
Thanks, that works now!  Could you also point me to the client sandboxes (and open them up), please?

**Comment by lee212** [comment]
I don't have the client sandboxes, let me ask Eugen if he keeps them.

**Comment by andre-merzky** [comment]
Without the client sandboxes it's hard to tell what's going on.  The tasks get submitted late - the first one arrives 48 minutes after the pilot becomes active, so it looks like RE is delaying submission, or RP spends time on the client side.  The tasks get running within 2 seconds after arrival, so that is not the problem...

**Comment by lee212** [comment]
Okay, I was told that the client sandbox are not kept anymore. Now, I guess the only way to find out what happened is to replicate the same configuration and see how it performs.

**Comment by andre-merzky** [comment]
Do we still have the session structure in the database?  Or the task and pilot descriptions?

**Comment by mturilli** [comment]
We are missing the client side of the sandbox and this issue might have depended on NFS issues. This is now low priority waiting for opportunity to replicate the issue.

**Comment by lee212** [comment]
I managed to run the workflow with the results:
- node counts: replicas / 6
- units: elapsed time in seconds from the beginning at 0
- umgr scheduling: entk starts to inject tasks
- agent executing: point of the first task
- last task: point of the last task

| replicas  | trial | pmgr launching | agent executing | umgr scheduling | agent executing | last task |
| ------------- | ------------- | ------------- | ------------- | ------------- | ------------- | ------------- |
| 60  | 1st  | 11.02 | 90.05 | 108.50 | 144.59 | 154.00 |
| 60 | 2nd | 12.20 | 91.95 | 110.76 | 148.29 | 154.00 |
| 60 | 3rd | 10.24 | 401.61 | 419.68 | 457.96 | 463.35 |
| 120 | 1st | 10.92 | 1743.62 | 1758.61 | 1820.55 | 1833.11 |
| 120 | 2nd | 11.28 | 237.03 | 250.59 | 313.93 | 325.90 |
| 120 | 3rd | 11.13 | 97.22 | 110.83 | 175.08 | 187.39 |
| 240 | 1st | 11.96 | 1733.27 | 1754.02 | 1869.44 | 1894.42 |
| 240 | 2nd | 10.96 | 1290.05 | 1307.79 | 1422.17 | 1447.61 |
| 240 | 3rd | 8.94 | 91.05 | 109.32 | 222.65 | 251.44 |

Base path of session data: `$WORLDWORK/csc393/hrlee/gpu_scaling_test/`

| replicas | trial | session id |
| -------- | ----- | ---------- |
| 60 | 1st | re.session.login4.hrlee.018404.0005 | 
| 60 | 2nd | re.session.login4.hrlee.018404.0006 |
| 60 | 3rd |  re.session.login4.hrlee.018404.0007 |
| 120 | 1st | re.session.login4.hrlee.018404.0008 |
| 120 | 2nd | re.session.login4.hrlee.018404.0009 |
| 120 | 3rd |  re.session.login4.hrlee.018404.0010 |
| 240 | 1st | re.session.login3.hrlee.018404.0020 |
| 240 | 2nd |  re.session.login3.hrlee.018404.0021 | 
| 240 | 3rd | re.session.login3.hrlee.018404.0022 |

**Comment by mturilli** [comment]
Based on what discussed, the overheads we want to look at it:


replicas | Mean overhead | STD
-- | -- | --
60 | 37.30 | 1.11
120 | 63.17 | 1.16
240 | 114.37 | 1.04

Approximating, we can say that the overhead ~doubles with the number of replicas (i.e., of GPUs as 1 replica = 1 GPU).  If this behavior is maintained at scale (only way to know for sure would be to run an actual test), we should end up with ~30m overheads at ~4K replicas/GPUs. This is way off what reported. 

I would proceed with our analysis, looking at the dominant overheads between `umgr scheduling` and `agent executing` in the RP profiles of the session `re.session.login3.hrlee.018404.0022`. Note that that session uses JSRUN so it would be interesting to have another one that uses PRRTE for comparison. 

Happy to help with the analysis if helpful. Just let me know and I will write an analysis on a jupyter notebook and I will share it.

**Comment by lee212** [comment]
The status of the scaling test is finished up to 2k GPU tasks, the results are available here: https://docs.google.com/spreadsheets/u/2/d/1gwbiQ2ngV8XQYuBsl9aOOP0TKHvmsGZdeLoAxqVIVnw/edit#gid=0

4k will be conducted soon with multi agents.

**Comment by mturilli** [comment]
Multi agent support on Summit is now fixed. The activity has been tracked here https://github.com/radical-cybertools/radical.pilot/issues/2186

**Comment by mturilli** [comment]
Multi agent support has been fixed and released in 1.5.

**Comment by lee212** [comment]
I propose to close this as we achieved initial goals.
